### ISSUE TITLE
Fyne app's icon always looks bad

### ISSUE THREAD

**Issue by Johnnycyan** [fyne-io/fyne#4109]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

I should preface this is probably only an issue on Windows.

The main problem here is that no matter what size image I use for an icon it always will look bad in at least 1 scenario. Whether it be that it looks bad in the program's title bar, it looks bad in the Start Menu, or it looks bad in the Taskbar. This is because Fyne's packaging seems to only support 1 icon size. Whereas most applications actually use a .ico package that contains sizes of 16, 24, 32, 48, 64, 128, and 256 pixels. having all these options means the program doesn't have to apply scaling to the image which creates bad aliasing artifacts. I can confirm this is the case by opening a Fyne packaged .exe with Resource Hacker and see only the 1 icon size. This is even the case if you give the .toml file an .ico package with multiple sizes. Whereas opening any other normal application with Resource Hacker results in seeing all of those icon sizes. This issue is actually fixed if you manually create a .ico package and replace the existing icon inside the .exe with Resource Hacker. But Fyne's Packager should by default create these multi-sized icon packages automatically or at the very least let me supply an already created one and actually respect it.

### How to reproduce

1. Package a go file including an icon with Fyne for Windows e.g. `fyne package -os windows --release true`
2. Observe how the icon looks for various locations in Windows (Taskbar, Start Menu, File Browser etc.)

### Screenshots

Example with supplying the Fyne packager with a multi-sized .ico resulting in a single 16px icon in the .exe file.

Looks fine in file explorer:
![image](https://github.com/fyne-io/fyne/assets/24556317/890378a6-b45d-41a7-be61-b24520f3e30e)

Looks terrible in Start Menu:
![image](https://github.com/fyne-io/fyne/assets/24556317/d6eba80b-7638-48bb-ab31-33c6c76e4d39)

Looks terrible in Taskbar:
![image](https://github.com/fyne-io/fyne/assets/24556317/f1848ad3-be6d-419f-b941-dad2bb8c6f85)

Looks good in application title bar:
![image](https://github.com/fyne-io/fyne/assets/24556317/3e260a41-3d3d-4517-bce6-80b7289dfed3)

Example after replacing the icon with Resource Hacker.

Looks great in file explorer:
![image](https://github.com/fyne-io/fyne/assets/24556317/93dfee57-ebc9-41c1-b052-82a8ec0f1eae)

Looks great in Start Menu:
![image](https://github.com/fyne-io/fyne/assets/24556317/95132ced-49ca-47d3-ac1e-c2bfe513a91b)

Looks great in Taskbar:
![image](https://github.com/fyne-io/fyne/assets/24556317/8a7be6e8-5731-45ba-bbb8-59d0f8afb2e4)

Looks great in application title bar:
![image](https://github.com/fyne-io/fyne/assets/24556317/9f2d0998-f940-4319-a657-2a4d45faef26)

### Example code

Has nothing to do with code

### Fyne version

v2.3.6-0.20230717215520-4e63455578cc

### Go compiler version

1.20.6

### Operating system and version

Windows 11 22H2

**Comment by andydotxyz** [comment]
Generally we recommend 1024x1024px as your source - macOS/iOS require this for some uses. 

Updating to represent the feature request this seems to convey.

**Comment by Johnnycyan** [comment]
> Generally we recommend 1024x1024px as your source - macOS/iOS require this for some uses.
> 
> Updating to represent the feature request this seems to convey.

Using 1024 in Windows causes it to look bad in every place on windows due to it being scaled badly. And yeah sorry wasn't really sure where to put this

**Comment by Jacalz** [comment]
I've also seen some bad looking icons on Windows but also Linux and I think it is related to bad scaling. My 512x512 pixel icon looks very blocky and unsmooth sometimes. Will attach some screenshots tomorrow (if I remember to).

**Comment by navossoc** [comment]
I started using Fyne two days ago and I stumbled in the same issue.
Probably I will look into another way to embed the icons on the executable.

But I agree with @Johnnycyan said.
If it is possible to supply multiple images or icons this will fix the issue.

Like:
Icon.png = one image for all
Icon_512.png = if you want to override this specific size
Icon_32.png = same...

Something like that should work fine.
Just create some rules and good to go.

It won't break anything and if someone wants to use, it will be easier to use.

[]'s

**Comment by andydotxyz** [comment]
Maybe we should push towards SVG files that we can scale up/down for any required file sizes? We do that in the main UI and could do that for icons too I guess

**Comment by dweymouth** [comment]
I'm not sure if the original reporter was hoping for this but I know some apps actually have different designs for different icon resolutions, eg intentionally omitting some fine details in the 32x32 or 64x64 versions. If we want to eventually support that then a single SVG or hi-res png isn't enough.

**Comment by navossoc** [comment]
I agree with @dweymouth, different designs are more common. Sometimes you can't fit a whole image in 32x32.

I think you can leave as it is: one Icon.png and let the OS resize it, but you should expand it to allow people further customize suppling more images/icons.

**Comment by Johnnycyan** [comment]
Yeah for now I just create my own multi-sized icon file and use Resource Hacker to replace the icon resource in the compiled exe. Don't know about a linux alternative.

**Comment by Jacalz** [comment]
No discussions have taken place as far as I know. If there is something happening then it is usually indicated on the issue.

I have been doing this manually on Linux for a long time. The icon is exported from Inkscape in the right sizes and the Makefile installs them all: https://github.com/Jacalz/hegelmote/blob/main/Makefile

**Comment by adamk33n3r** [comment]
Hmmm, I tried using Resource Hacker to replace the icon. I replaced it with an ico file that has multiple different sizes. But when opening the app it's still showing the crunchy version in the titlebar. Actually, I just replaced the icon with a completely different icon and it's still showing the original one. So apparently the Resource Hacker method doesn't do anything and fyne overrides it at runtime?

**Comment by Johnnycyan** [comment]
> Hmmm, I tried using Resource Hacker to replace the icon. I replaced it with an ico file that has multiple different sizes. But when opening the app it's still showing the crunchy version in the titlebar. Actually, I just replaced the icon with a completely different icon and it's still showing the original one. So apparently the Resource Hacker method doesn't do anything and fyne overrides it at runtime? Looks like it updates the icon that is shown in Windows Explorer. But not the icons when the app is running in the titlebar or taskbar.

The taskbar does get updated, but windows caches icon files so you'd most likely have to restart your pc or invalidate the cache, I don't remember how that's done. But yes the app icon in-app doesn't get changed. I assume that's part of the program code itself somewhere

**Comment by adamk33n3r** [comment]
I've added support for ico files in my own forks. You can see the changes made [here](https://github.com/fyne-io/fyne/compare/develop...adamk33n3r:fyne:develop) and [here](https://github.com/fyne-io/tools/compare/main...adamk33n3r:fyne-tools:main). If you would like me to make some PRs I'll gladly do that. I'm guessing I'd need to add some tests but let me know. In the meantime, I can do a `replace` in my `go.mod` to use my fork.

**Comment by Jacalz** [comment]
.ico is very much a Windows specific (or at least only used there widely) so I'm not sure to whoah degree we'd want to support hat everywhere. Passing a .ico file as the main icon does not make much sense on Linux (and likely macOS as well). I suppose it would be better to scale a big PNG file down into sensible sizes and use that instead (maybe converting it to an .ico file on Windows if we have to)? The same approach should work on Linux

**Comment by adamk33n3r** [comment]
Ya I thought about ico being pretty Windows, but I figured it would be nice to support that as an option. I agree though for other platforms another solution would probably be better. Perhaps adding more properties to the toml for different sizes or what you suggest, resizing during build. You would definitely want a way for a dev to provide unique images per size as mentioned above that sometimes smaller icons have different design. So maybe you could let the dev supply multiple sizes and then fyne fills in the rest.

To answer your question, you probably wouldn't need to convert to ico for Windows as fyne (in my fork) is just reading in the ico file via `ico.DecodeAll` and getting a list of every image inside to pass to glfw. This could be replaced with however it is decided above. Though after saying that if what @Johnnycyan said is true and that it does indeed read the ico in the resources, maybe you still would.
